cursor.c

```diff
@@ -277,6 +285,18 @@ PHP_METHOD(MongoCursor, hasNext)
 	if (!php_mongo_get_more(cursor TSRMLS_CC)) {
 		RETURN_FALSE;
 	}
+
+	/* Another special case. Because we had tested for *one more* above, we do
+	 * need to check whether there are actually more results before we
+	 * conclusively can say there are more results on the cursor. We try to run
+	 * getMore here, and then check whether it had changed the "num" from the
+	 * last value. The last value of "num" is now "start", so that's why we
+	 * compare it with that. And of course, only if the cursor ID is now 0 (no
+	 * more results after this). */
+	if (cursor->cursor_id == 0 && cursor->start == cursor->num) {
+		RETURN_FALSE;
+	}
```

Comment:
> Is this only triggered by a getmore returning no results? Looking at
> php_mongo_get_cursor_header(), cursor->start is set when processing an
> OP_REPLY.
>
> When does a getmore return no results (assuming we're not dealing with a
> tailable cursor)? I would expect the previous OP_REPLY to report that the
> cursor has been closed if all results could fit in a response.

I would expect that too, but that's not what happens. The cursor->id is only set to 0 when getmore is called for the 2nd time. The first time being after the first 2 results, and the second time after 4 read results. The 2nd getmore returns 0 extra documents, and then sets the cursor ID to 0. See the new test case I have pushed to https://github.com/jmikola/mongo-php-driver/pull/7/files


> One of the first checks we do in hasNext() is 
> ((cursor->limit > 0 && cursor->at >= cursor->limit) || cursor->num == 0), 
> so I would expect the latter or conditional would catch an empty response
> (i.e. cursor->num == 0).

cursor->num being 0 only happens if there are *no results* at all. getmore doesn't change this number if it finds no more new results. It only increases it when getmore returns more results. So no, it wouldn't catch it if it's not the "first batch" that comes back with the first OP_REPLY.